File: embed/_voyage.py
from collections.abc import Sequence
from struct import pack

from voyageai.client import Client

_client = Client(max_retries=5)


def _encode(embedding: Sequence[float]) -> bytes:
    return pack("<1024f", *embedding)


def _embed_texts(texts: list[str]) -> list[bytes]:
    result = _client.embed(
        texts=texts,
        model="voyage-4-large",
        input_type="document",
        output_dimension=1024,
    )

    return [_encode(embedding) for embedding in result.embeddings]

File: packages/minecraft-corpus/src/voyage.ts
import { VoyageAIClient } from "voyageai";

const voyage = new VoyageAIClient({
  apiKey: process.env.VOYAGE_API_KEY as string,
});

export async function embedQuery(text: string): Promise<number[]> {
  const response = await voyage.embed({
    input: [text],
    model: "voyage-4-large",
    inputType: "query",
    outputDimension: 1024,
  });

  return response.data![0].embedding!;
}

export async function embedMultimodalQuery(text: string): Promise<number[]> {
  const response = await voyage.multimodalEmbed({
    inputs: [{ content: [{ type: "text", text }] }],
    model: "voyage-multimodal-3.5",
    inputType: "query",
  });

  return response.data![0].embedding!;
}

export async function rerank(
  query: string,
  documents: string[],
): Promise<{ index: number; score: number }[]> {
  const response = await voyage.rerank({
    query,
    documents,
    model: "rerank-2.5-lite",
    topK: documents.length,
  });

  return response.data!.map((result) => ({
    index: result.index!,
    score: result.relevanceScore!,
  }));
}
